apps/rush-lib/src/RushX.ts

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as child_process from 'child_process';
+import * as colors from 'colors';
+import * as os from 'os';
+import * as path from 'path';
+
+import {
+  PackageJsonLookup,
+  IPackageJson,
+  Text,
+  IPackageJsonScriptTable
+ } from '@microsoft/node-core-library';
+import { Utilities } from '../lib/utilities/Utilities';
+
+export class RushX {
+
+  public static launchRushX(launcherVersion: string, isManaged: boolean): void {
+    // NodeJS can sometimes accidentally terminate with a zero exit code  (e.g. for an uncaught
+    // promise exception), so we start with the assumption that the exit code is 1
+    // and set it to 0 only on success.
+    process.exitCode = 1;
+
+    try {
+      // Find the governing package.json for this folder:
+      const packageJsonLookup: PackageJsonLookup = new PackageJsonLookup();
+
+      const packageJsonFilePath: string | undefined = packageJsonLookup.tryGetPackageJsonFilePathFor(process.cwd());
+      if (!packageJsonFilePath) {
+        console.log(colors.red('This command should be used inside a project folder.'));
+        console.log(`Unable to find a package.json file in the current working directory or any of its parents.`);
+        return;
+      }
+
+      const packageJson: IPackageJson = packageJsonLookup.loadPackageJson(packageJsonFilePath);
+
+      // 0 = node.exe
+      // 1 = rushx
+      const args: string[] = process.argv.slice(2);
+
+      if (args.length === 0 || args[0][0] === '-') {
+        RushX._showUsage(packageJson);
+        return;
+      }
+
+      const command: string = args[0];
+
+      const scripts: IPackageJsonScriptTable = packageJson.scripts || { };
+
+      if (!Object.hasOwnProperty.call(scripts, command)) {
+        console.log(colors.red(`Error: The command "${command}" is not defined in the`
+          + ` package.json file for this project.`));
+
+        const availableCommands: string[] = Object.keys(scripts);
+        if (availableCommands.length > 0) {
+          console.log(os.EOL + 'Available commands for this project are: '
+            + availableCommands.map(x => `"${x}"`).join(', '));
+        }
+
+        console.log(`Use ${colors.yellow('"rushx --help"')} for more information.`);
+        return;
+      }
+
+      const scriptBody: string = scripts[command];
+      console.log('Executing: ' + JSON.stringify(scriptBody) + os.EOL);
+
+      const packageFolder: string = path.dirname(packageJsonFilePath);
+
+      const result: child_process.SpawnSyncReturns<Buffer> = Utilities.executeLifecycleCommand(
+        scriptBody,
+        packageFolder,
+        packageFolder
```

Comment:
We should consider making this an options object.

Reply:
Yes but that's out of scope for this PR.  The whole executeLifecycleCommand() is pretty rudimentary right now and should get overhauled

---
In reply to: [193581035](https://github.com/Microsoft/web-build-tools/pull/678#discussion_r193581035) [](ancestors = 193581035)